posts/simp-made-simple.md

```diff
@@ -0,0 +1,343 @@
+
+## `Step`
+
+[`Step`](https://leanprover-community.github.io/mathlib4_docs/find/?pattern=Lean.Meta.Simp.Step#doc) is the type that represents a single step in the simplification loop. In `simp`'s algorithm, a step intuitively corresponds to two pieces of information:
+1) The **result** of simplifying an expression `e`,
+2) The **location** of what should be simplified next, and in which **direction** (pre or post).
+
+The result of simplifying `e` is encoded as an expression `e'` and a proof that `e = e'`.
+This is encapsulated by the [`Result`](https://leanprover-community.github.io/mathlib4_docs/find/?pattern=Lean.Meta.Simp.Result#doc)
+structure:
+```lean
+structure Result where
+  /-- The new expression `e'` -/
+  expr   : Expr
+  /-- The proof that `e = e'` -/
+  proof? : Option Expr := none
+  -- Note, `Result` currently has an extra `cache` field that is both deprecated
+  -- and irrelevant to our current discussion.
+```
+> The proof is optional:
+> If `proof?` is set to its default `none` value, the equality is assumed to be definitional.
+
+> The proof is allowed to be an arbitrary `Expr`,
+> i.e. nothing ensures that it actually is a proof that `e = e'`.
+> It is up to the author of the procedure to make sure that the generated proof terms are valid.
+
+After simplifying the current expression `e` to a new expression `e'`,
+there are a few possible options for the next location:
+1. Simplify `e'` further.
+  Preprocedures are tried on `e'`.
+2. Simplify subexpressions of `e'`.
+  Preprocedures are tried on each child expression of `e'` in turn.
+  If `e'` has no child, then we run postprocedures on `e'`.
+3. Don't simplify further.
+  Preprocedures are tried on the next child of the parent expression.
+  If there is no such child, then postprocedures are tried on the parent expression.
+
+The three possibilities above correspond to the three constructors of `Step`:
+```lean
+inductive Step where
+  /-- Try preprocedures on the simplified expression. -/
+  | visit (e : Result)
+  /--
+  For `pre` procedures, continue transformation by visiting subexpressions, and then
+  executing `post` procedures.
+
+  For `post` procedures, this is equivalent to returning `visit`.
+  -/
+  | continue (e? : Option Result := none)
+  /-- Returns the result without visiting any subexpressions. -/
+  | done (r : Result)
+-- Note: the docstrings here are simplified versions of the real docstrings (which can be found in the Lean source code)
+```
+> If a procedure fails to simplify an expression, it should return `continue none`.
+  Both `visit` and `done` signify success.
+
+Whenever a simproc is called on a given expression, it outputs a `Step`, which determines what will happen next during the `simp` call. 
+Since every simproc call is running a metaprogram to produce the output `Step`, the constructor that ends up used may vary according to the input. 
+For example, a given simproc may in some cases use `visit` and in others use `continue`.
```

Comment:
This sentence is quite strange to me: it bases itself off the assumption that people expect that a simproc always uses the same `Step` constructor, but four lines above we explain how to use the constructors according to whether the simproc has failed or not. Could you repackage the info in a way that acknowledges the previous sentence?